File: medium/cpp/1372.cpp
// https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int longestZigZag(TreeNode* root)
    {
        auto res = 0;
        static_cast<void>(dfs(root, res));
        return res;
    }

private:
    std::pair<int, int> dfs(TreeNode* root, int& r)
    {
        if (root == nullptr)
        {
            return { 0, 0 };
        }

        auto res = std::make_pair(0, 0);

        if (root->left != nullptr)
        {
            const auto left = dfs(root->left, r);
            res.first = 1 + left.second;
        }

        if (root->right != nullptr)
        {
            const auto right = dfs(root->right, r);
            res.second = 1 + right.first;
        }

        r = std::max(r, std::max(res.first, res.second));

        return res;
    }
};

File: medium/python/1372.py
# https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(root):
            if root is None:
                return [0, 0]

            lens = [0, 0]

            if root.left is not None:
                sub_lens = dfs(root.left)
                lens[0] = 1 + sub_lens[-1]
            if root.right is not None:
                sub_lens = dfs(root.right)
                lens[-1] = 1 + sub_lens[0]

            nonlocal res
            res = max(res, lens[0], lens[-1])

            return lens

        dfs(root)

        return res
        